### DISCUSSION TITLE
Interleaved output when tasks are executed in parallel

### DISCUSSION THREAD

**Discussion by serpro69** [jdx/mise#3428] [Q&A]
When I run `mise run --jobs 1 sometask`, the output is OK, but when I try to execute jobs in parallel, I'm getting a lot of poorly-formatted output

![image](https://github.com/user-attachments/assets/5b973658-f06a-4248-a7f0-49e798c61fac)

I've tried a bunch of things, but nothing seems to fix this. I thought that mise was supposed to handle longer outputs with parallel jobs, but doesn't seem like it? (or I'm doing something wrong here...)

Any tips on how to avoid this behavior and being able to run tasks in parallel without messing out the output?

**Comment by shousper** [comment]
Just came across this problem, here is a little example I whipped up to demonstrate:

```
# Interleaved output, difficult to reconcile even if your commands prefix all outputs.
[tasks."test"]
run = { tasks = ["test:*"] }

[tasks."test:a"]
run = """
echo "task a: start" >&2; for x in $(seq 10); do echo "task a $x"; sleep 1; done; echo "task a: done" >&2
"""

[tasks."test:b"]
run = """
echo "task b: start" >&2; for x in $(seq 10); do echo "task b $x"; sleep 1; done; echo "task b: done" >&2
"""

# Buffered outputs
[tasks."test-buffered"]
run = { tasks = ["test-buffered:*"] }

[tasks."test-buffered:a"]
run = """
OUTPUT=$(mktemp)
trap 'rm $OUTPUT' EXIT
mise run test:a >$OUTPUT 2>&1
cat $OUTPUT
"""

[tasks."test-buffered:b"]
run = """
OUTPUT=$(mktemp)
trap 'rm $OUTPUT' EXIT
mise run test:b >$OUTPUT 2>&1
cat $OUTPUT
"""
```

It's not convenient to have to wrap any command that could be run in parallel like this. Maybe if the parallel task configuration just supported a `output = "buffered"` mode?

Otherwise.. maybe some kind of task "wrapping" to execute tasks using another task, e.g.:

```
[tasks."test"]
run = { tasks = ["blah:*"], via = "buffer-out" }

[tasks."buffer-out"]
run = """
OUTPUT=$(mktemp)
trap 'rm $OUTPUT' EXIT
cat /dev/stdin >$OUTPUT
cat $OUTPUT
"""
```

**Reply by jdx** [reply]
idk there used to be some hacky code that tries to reformat the output automatically. I took that out when I added support for directly defining tasks in toml. That said, it doesn't account for this use-case where you want to dynamically launch tasks.

I haven't fleshed this idea out fully, but I was thinking something like mise could start up a JSON-RPC server that could be used by sub-tasks to enqueue tasks back to the parent. I think that'd be the best way to avoid any issues with regards to output since only 1 mise process would actually be launching the tasks–it'd also have performance benefits since you don't need to pay the overhead of setting up the mise env each time.

**Reply by serpro69** [reply]
Thanks for your input @shousper , at least I know I'm not alone in this :grin: 
My use-case is exactly the same, I have a bunch of similar tasks, and then I have another "aggregate task" that runs all of them dynamically via `*`, similar to your `run = { tasks = ["test:*"] }` example. I have not found any good way to handle this issue and ended up writing output to a file myself. Not ideal but it works at least... otherwise the output was pretty unusable for me.

**Reply by shousper** [reply]
I was actually on the verge of writing my own task executor (also in rust) when someone introduced me to mise -- now I'm all in 💪🏻

---

My primary real world use case is spawning a boat load of `tofu init` processes to setup a monorepo (amongst some other bits).

In the end, not a big deal once I figured out how to use `-o prefix` on an `init-all` task, while inner tasks that are called use `mise run -o interleave ..` -- but I DID run into 1 scenario where it seemed to lose the `-o interleave`, which I fixed with an `export MISE_TASK_OUTPUT=interleave`. Haven't figured out what went wrong yet, so haven't raised an issue. `-o prefix` works well enough though 👍🏻

```toml
[tasks."tofu:init-all"]
description = "Initialize all OpenTofu configurations in parallel"
run = "mise run -o prefix 'tofu-init:*'"
```

I am however using _another_ setup task to _generate_ tasks for each tofu configuration into `./.mise/config.toml` like this:

```toml
[tasks."tofu-init:01-foundation-01-state"]
description = "Initialize 01-foundation/01-state"
dir = "{{config_root}}/01-foundation/01-state"
hide = true
run = "mise run tofu:init"
```

Maybe that'll help someone else scratch the same itch I had.

---

An RPC server is a pretty fancy idea 🤔 I was fine with predetermining the tasks to be run, then offloading the set to an executor. I'd be curious how you expect users to push tasks back into mise via RPC. curl or whatever? or a specific API client made available to the task process?

I played around with crossterm and indicatif to get different sorts of summarised, live updating output for parallel tasks, but it's really not fun. If you want stdout+stderr to retain their exact output order, you need a pty or file piping/redirection 🤮 For example, even trying to live-stream the last 5 lines of output from 4 tasks in a nice way seemed super fragile to me. Super disruptive to regular i/o.

Would be cool to be able to execute interactive tasks in parallel though. `mise run -E test tofu:apply` my whole env, review each config with changes, selectively accepting 🔥 That was my dream anyway 😅

**Reply by jdx** [reply]
> An RPC server is a pretty fancy idea 🤔 I was fine with predetermining the tasks to be run, then offloading the set to an executor. I'd be curious how you expect users to push tasks back into mise via RPC. curl or whatever? or a specific API client made available to the task process?

they wouldn't know the RPC server is there. We'd set an environment variable like `MISE_TASK_RPC=/tmp/mise-task-run.json` that points to a unix socket and if that is set it, then `mise run` would communicate with the RPC server rather than running any tasks.

I think something like that is the only way for the output to work properly since we can't control how users launch tasks inside a task.

**Reply by jdx** [reply]
> Would be cool to be able to execute interactive tasks in parallel though

I think this is impossible

**Reply by shousper** [reply]
> then `mise run` would communicate with the RPC server

Mint 👌🏻 If you're thinking to pipe process outputs through IPC without spinning up a shell, you'll probably need separate sockets for each process, right? Definitely you'll have a lot more output control, but it'll definitely be a lot harder to support interactive tasks.

> I think this is impossible

I was hoping if you can hand the executor any stream for stdin/out/err, then you can manipulate them as you please. Detecting when a process is waiting for input, so you can bring that prompt forward to the user is far from simple though. A task could detect specific patterns in the output, but it's fragile. The experience becomes more like a TUI in that case though =/

**Comment by JamBalaya56562** [comment]
This is a setting, and it landed twelve days after this was posted — which is why nothing here mentions it.

`task.output = "keep-order"` buffers each task's output and prints it in one contiguous block. Measured on **v2026.8.3** with two tasks that each print 5 lines with a `sleep 0.15` between them:

```toml
[tasks."test:a"]
run = 'for i in 1 2 3 4 5; do echo "task a line $i"; sleep 0.15; done'
[tasks."test:b"]
run = 'for i in 1 2 3 4 5; do echo "task b line $i"; sleep 0.15; done'
[tasks.all]
depends = ["test:a", "test:b"]
```

```console
$ mise run all                                  # default (prefix)
[test:a] task a line 1
[test:b] task b line 1
[test:a] task a line 2
[test:b] task b line 2
[test:a] task a line 3
[test:b] task b line 3
[test:a] task a line 4
[test:b] task b line 4
[test:a] task a line 5
[test:b] task b line 5
[test:a] Finished in 785.4ms
[test:b] Finished in 772.4ms
Finished in 888.3ms

$ MISE_TASK_OUTPUT=keep-order mise run all
[test:a] task a line 1
[test:a] task a line 2
[test:a] task a line 3
[test:a] task a line 4
[test:a] task a line 5
[test:a] Finished in 770.8ms
[test:b] task b line 1
[test:b] task b line 2
[test:b] task b line 3
[test:b] task b line 4
[test:b] task b line 5
[test:b] Finished in 770.1ms
Finished in 850.3ms
```

**The tasks still run in parallel** — 850.3ms against 888.3ms for two ~770ms tasks. Only the display is buffered.

Three ways to set it:

```toml
[settings]
task.output = "keep-order"
```

`MISE_TASK_OUTPUT=keep-order`, or `mise run -o keep-order`. (The flat `task_output` key still works but is deprecated in favour of `task.output`.)

The full set of values today is `prefix` (the default), `keep-order`, `interleave`, `replacing`, `timed`, `quiet` and `silent`. `interleave` is the opposite of what you want here — it drops the prefixes and lets the streams mix.

`keep-order` was added in #3763 and first released in **v2024.12.17** (2024-12-21), alongside `replacing` (#3764) and `timed` (#3766). Bisected on the source rather than taken from the changelog: `src/cli/run.rs` has no `keep-order` at v2024.12.16 and has it at v2024.12.17.

@shousper — for the aggregate `run = { tasks = ["test:*"] }` shape in your example this is the one to reach for.
